datavec/datavec-python/src/main/java/org/datavec/python/PythonObject.java

```diff
@@ -77,78 +82,69 @@ public PythonObject(BytePointer bp){
-                ctype = PyObject_GetAttrString(ctypes, "c_uint32");
+                numpyType = NPY_UINT;
                 break;
             case UINT64:
-                ctype = PyObject_GetAttrString(ctypes, "c_uint64");
+                numpyType = NPY_UINT64;
                 break;
             case BOOL:
-                ctype = PyObject_GetAttrString(ctypes, "c_bool");
+                numpyType = NPY_BOOL;
                 break;
             case BYTE:
-                ctype = PyObject_GetAttrString(ctypes, "c_byte");
+                numpyType = NPY_BYTE;
                 break;
             case UBYTE:
-                ctype = PyObject_GetAttrString(ctypes, "c_ubyte");
+                numpyType = NPY_UBYTE;
+                break;
+            case HALF:
+                numpyType = NPY_HALF;
                 break;
             default:
                 throw new RuntimeException("Unsupported dtype: " + npArray.getDtype());
         }
 
-        PyObject ctypesPointer = PyObject_GetAttrString(ctypes, "POINTER");
-        PyObject argsTuple = PyTuple_New(1);
-        PyTuple_SetItem(argsTuple, 0, ctype);
-        PyObject ptrType = PyObject_Call(ctypesPointer, argsTuple, null);
-
-        PyObject cast = PyObject_GetAttrString(ctypes, "cast");
-        PyObject address = PyLong_FromLong(npArray.getAddress());
-        PyObject argsTuple2 = PyTuple_New(2);
-        PyTuple_SetItem(argsTuple2, 0, address);
-        PyTuple_SetItem(argsTuple2, 1, ptrType);
-        PyObject ptr = PyObject_Call(cast, argsTuple2, null);
-        PyObject shapeTuple = PyTuple_New(npArray.getShape().length);
-        for (int i = 0; i < npArray.getShape().length; i++) {
-            PyObject dim = PyLong_FromLong(npArray.getShape()[i]);
-            PyTuple_SetItem(shapeTuple, i, dim);
-            Py_DecRef(dim);
+        long[] shape = indArray.shape();
+        INDArray inputArray = indArray;
+        if(dataType == DataType.BFLOAT16) {
+            log.warn("\n\nThe given nd4j array \n\n{}\n\n is of BFLOAT16 datatype. " +
+                    "Casting a copy of it to FLOAT and creating the respective numpy array from it.\n", indArray);
+            inputArray = indArray.castTo(DataType.FLOAT);
         }
-        PyObject ctypesLib = PyObject_GetAttrString(np, "ctypeslib");
-        PyObject asArray = PyObject_GetAttrString(ctypesLib, "as_array");
-        PyObject argsTuple3 = PyTuple_New(2);
-        PyTuple_SetItem(argsTuple3, 0, ptr);
-        PyTuple_SetItem(argsTuple3, 1, shapeTuple);
-        nativePythonObject = PyObject_Call(asArray, argsTuple3, null);
 
-        Py_DecRef(ctypesPointer);
-        Py_DecRef(ctypesLib);
-        Py_DecRef(argsTuple);
-        Py_DecRef(argsTuple2);
-        Py_DecRef(argsTuple3);
-        Py_DecRef(cast);
-        Py_DecRef(asArray);
+        //Sync to host memory in the case of CUDA, before passing the host memory pointer to Python
+        if(inputArray.data() instanceof BaseDataBuffer){
```

Comment:
@AlexDBlack If this check is required for CUDA only, perhaps `if (inputArray.data() instaceof BaseCudaDataBuffer)` is better?

Reply:
syncToPrimary is a no-op on CPU
and we can't do `instanceof BaseCudaDataBuffer` as that requires a cuda dependency